// Generated from service_discovery.ifex by covesa-ifex-core
// DO NOT EDIT - regenerate with: ./generate_proto.sh
syntax = "proto3";



package swdv.service_discovery;

// Namespace: registry
// Service registration and discovery functionality

// Current status of a registered service
enum service_status_t {
  AVAILABLE = 0;
  UNAVAILABLE = 1;
  STARTING = 2;
  STOPPING = 3;
  ERROR = 4;
}

// Transport protocol used by service
enum transport_type_t {
  GRPC = 0;
  DBUS = 1;
  SOMEIP = 2;
  HTTP_REST = 3;
  MQTT = 4;
}

// IFEX parameter data types
enum parameter_type_t {
  UINT8 = 0;
  UINT16 = 1;
  UINT32 = 2;
  UINT64 = 3;
  INT8 = 4;
  INT16 = 5;
  INT32 = 6;
  INT64 = 7;
  FLOAT = 8;
  DOUBLE = 9;
  BOOLEAN = 10;
  STRING = 11;
  BYTES = 12;
  STRUCT = 13;
  ARRAY = 14;
  ENUM = 15;
}

// Information about a method parameter
message parameter_info_t {
  string name = 1;
  parameter_type_t type = 2;
  bool is_optional = 3;
  string description = 4;
  string min_value = 5;
  string max_value = 6;
  repeated string enum_values = 7;
}

// Information about a service method
message method_info_t {
  string name = 1;
  string description = 2;
  repeated parameter_info_t input_parameters = 3;
  repeated parameter_info_t output_parameters = 4;
  uint32 max_execution_time_ms = 5;
}

// Information about a service namespace
message namespace_info_t {
  string name = 1;
  string description = 2;
  repeated method_info_t methods = 3;
}

// Service endpoint information
message service_endpoint_t {
  transport_type_t transport = 1;
  string address = 2;
  string path = 3;
  string security_info = 4;
}

// Complete service information
message service_info_t {
  string name = 1;
  string version = 2;
  string description = 3;
  service_status_t status = 4;
  service_endpoint_t endpoint = 5;
  repeated namespace_info_t namespaces = 6;
  string ifex_schema = 7;
  uint64 last_heartbeat = 8;
}

// Filter criteria for service queries
message service_filter_t {
  string name_pattern = 1;
  string has_method = 2;
  transport_type_t transport_type = 3;
  bool available_only = 4;
  repeated string extension_paths = 5;
}

// Method: register_service
// Register a service with the discovery system
message register_service_request {
  service_info_t service_info = 1;
}

message register_service_response {
  string registration_id = 1;
}

service register_service_service {
  rpc register_service(register_service_request) returns (register_service_response);
}

// Method: unregister_service
// Unregister a service
message unregister_service_request {
  string registration_id = 1;
}

message unregister_service_response {
}

service unregister_service_service {
  rpc unregister_service(unregister_service_request) returns (unregister_service_response);
}

// Method: get_service
// Get a specific service by name
message get_service_request {
  string service_name = 1;
}

message get_service_response {
  service_info_t service_info = 1;
}

service get_service_service {
  rpc get_service(get_service_request) returns (get_service_response);
}

// Method: query_services
// Query services matching filter criteria
message query_services_request {
  service_filter_t filter = 1;
}

message query_services_response {
  repeated service_info_t services = 1;
}

service query_services_service {
  rpc query_services(query_services_request) returns (query_services_response);
}

// Method: validate_method_call
// Validate if a method call is valid
message validate_method_call_request {
  string service_name = 1;
  string method_name = 2;
  string parameters_json = 3;
}

message validate_method_call_response {
  bool is_valid = 1;
  repeated string validation_errors = 2;
}

service validate_method_call_service {
  rpc validate_method_call(validate_method_call_request) returns (validate_method_call_response);
}

// Method: get_method_schema
// Get parameter schema for a method
message get_method_schema_request {
  string service_name = 1;
  string method_name = 2;
}

message get_method_schema_response {
  method_info_t method_info = 1;
}

service get_method_schema_service {
  rpc get_method_schema(get_method_schema_request) returns (get_method_schema_response);
}

// Method: heartbeat
// Service heartbeat to indicate availability
message heartbeat_request {
  string registration_id = 1;
  service_status_t status = 2;
}

message heartbeat_response {
}

service heartbeat_service {
  rpc heartbeat(heartbeat_request) returns (heartbeat_response);
}


